### PR TITLE
Ledger Top-Up Redesign

### LINKED ISSUE #543: Ledger Top Up protocol relies on duplicate addresses
Currently the Ledger Top up protocol is designed around adding duplicate addresses to the the destination on the `commitment`. However it's not clear that the nitro protocol fully supports this.

We should update it so it just re-orders the the destination/allocation instead of appending duplicate values. This means another round trip of `commitment`s but will make sure it works ok with the `nitro-protocol`

### PR DESCRIPTION
**Pull request by lalexgap**
This changes the Ledger Top-up protocol to no longer rely on using repeat addresses on the ledger channel.

It now relies on moving the player's allocation to the end (so they cannot withdraw more than their share) before depositing, following the algorithm outlined in the paper.

The tests are also more extensive and check that the correct amounts are used for the deposit and then ledger updates.

There are also some updates to DirectFunding to handle the case where we're depositing into a channel that has already undergone a round of funding.

Fixes #543 and #566

### CODE REVIEW

**Comment by geoknee** [comment]
Do you think the UI should be updated as part of this PR?
Screenshot 2019-06-28 at 13 53 05

**Comment by geoknee** [comment]
It would be great if you could update the protocol hierarchy diagram in [/notes/index.md](https://github.com/magmo/apps/blob/71c1e9ea445e8a7651999cb823297d07576e32d7/packages/wallet/notes/index.md) (probably after merging or rebasing off master).

File: packages/wallet/src/redux/protocols/ledger-top-up/readme.md [hunk 1]
```diff
@@ -2,34 +2,46 @@
...
 ## Scenarios
 
-1. **Player A Happy Path** Start->WaitForPreTopUpLedgerUpdate->WaitForDirectFunding->WaitForPostTopUpLedgerUpdate->Success
-2. **Player B Happy Path** Start->WaitForPreTopUpLedgerUpdate->WaitForDirectFunding->WaitForPostTopUpLedgerUpdate->Success
-3. **Pre-TopUp Invalid Update** WaitForPreTopUpLedgerUpdate->Failure
-4. **Post-TopUp Invalid Update** WaitForPostTopUpLedgerUpdate->Failure
-
-TODO: Should we have a scenario for cases where only one player needs to top up?
+1. **Player A Happy Path** Start->SwitchOrderAndAddATopUpUpdate->WaitForDirectFundingForPlayerA->RestoreOrderAndAddBTopUpUpdate->WaitForDirectFundingForPlayerB->Success
+2. **Player B Happy Path** Start->SwitchOrderAndAddATopUpUpdate->WaitForDirectFundingForPlayerA->RestoreOrderAndAddBTopUpUpdate->WaitForDirectFundingForPlayerB->Success
```
Comment (geoknee) [inline packages/wallet/src/redux/protocols/ledger-top-up/readme.md]:
```suggestion
2. **Player B Happy Path** Start->SwitchOrderAndAddATopUpUpdate->WaitForDirectFundingForPlayerA->RestoreOrderAndAddBTopUpUpdate->WaitForDirectFundingForPlayerB->Success
3. **(Player A) One user needs top-up** Start->SwitchOrderAndAddATopUpUpdate->WaitForDirectFundingForPlayerA->RestoreOrderAndAddBTopUpUpdate->Success
4. **(Player B) One user needs top-up** Start->SwitchOrderAndAddATopUpUpdate->WaitForDirectFundingForPlayerA->RestoreOrderAndAddBTopUpUpdate->Success
```

File: packages/wallet/src/redux/protocols/ledger-top-up/__tests__/reducer.test.ts [hunk 2]
```diff
@@ -2,9 +2,139 @@ import * as scenarios from './scenarios';
...
+    });
+  });
+  describeScenarioStep(scenario.waitForDirectFundingForB, () => {
+    const { action, sharedData, state } = scenario.waitForDirectFundingForB;
+    const updatedState = ledgerTopUpReducer(state, sharedData, action);
+    itTransitionsTo(updatedState, 'LedgerTopUp.Success');
+  });
+});
+
+describe('player A one use needs top up', () => {
```
Comment (geoknee) [inline packages/wallet/src/redux/protocols/ledger-top-up/__tests__/reducer.test.ts]:
```suggestion
describe('player A one user needs top up', () => {
```

File: packages/wallet/src/redux/protocols/ledger-top-up/__tests__/reducer.test.ts [hunk 3]
```diff
@@ -13,37 +143,53 @@ describe('player A both players need a top up', () => {
...
+  describeScenarioStep(scenario.restoreOrderAndAddBTopUpUpdate, () => {
+    const { action, sharedData, state } = scenario.restoreOrderAndAddBTopUpUpdate;
     const updatedState = ledgerTopUpReducer(state, sharedData, action);
     itTransitionsTo(updatedState, 'LedgerTopUp.Success');
   });
 });
 
-describe('player B both players need a top up', () => {
-  const scenario = scenarios.playerBBothPlayersTopUp;
+describe('player B one use needs top up', () => {
```
Comment (geoknee) [inline packages/wallet/src/redux/protocols/ledger-top-up/__tests__/reducer.test.ts]:
```suggestion
describe('player B one user needs top up', () => {
```

**Comment by lalexgap** [comment]
> Do you think the UI should be updated as part of this PR?
> Screenshot 2019-06-28 at 13 53 05

I've created #570 so we can get this merged in.

**Comment by lalexgap** [comment]
> It would be great if you could update the protocol hierarchy diagram in [/notes/index.md](https://github.com/magmo/apps/blob/71c1e9ea445e8a7651999cb823297d07576e32d7/packages/wallet/notes/index.md) (probably after merging or rebasing off master).

done!

**Pull request merged**